### PR TITLE
test: add BBR path-wins auth and streaming inference tests

### PR DESCRIPTION
**Pull request by SB159**
# Pull Request

## Summary

add BBR path-wins auth and streaming inference tests

## Related Issues

- Fixes: 
- JIRA: 

## Please review and indicate how it has been tested

- [x] Locally
- [ ] Jenkins

## Additional Requirements

- [ ] If this PR introduces a new test image, did you create a PR to mirror it in disconnected environment?
- [ ] If this PR introduces new marker(s)/adds a new component, was relevant ticket created to update relevant Jenkins job?

## Summary by CodeRabbit

* **Tests**
  * Added coverage for inference requests that fail authentication, including invalid or missing API keys.
  * Added a regression check to ensure route-based behavior is preserved even when the request body references a different model.
  * Added streaming coverage to verify inference responses return server-sent events with streamed data chunks.
  * Updated test setup to use module-level fixture application for cleaner coverage setup.

### CODE REVIEW

File: tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py [hunk 1]
```diff
@@ -63,6 +63,66 @@ def test_inference_model_in_body_only_invalid_key_returns_401(
             expected_status=401,
         )
 
+    @pytest.mark.tier1
+    @pytest.mark.parametrize("ocp_token_for_actor", [{"type": "free"}], indirect=True)
```
Comment (dbasunag) [inline tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py]: This is a class fixture, being used in function context. This does not seem right.

File: tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py [hunk 2]
```diff
@@ -63,6 +63,66 @@ def test_inference_model_in_body_only_invalid_key_returns_401(
...
+            timeout=60,
+        )
+        assert response.status_code not in (401, 403), (
+            f"Auth rejected based on body model; expected path identity to win on /llm/ paths, "
+            f"got {response.status_code}"
+        )
+        LOGGER.info(f"Path-wins auth verified: got {response.status_code} (not 401/403)")
+
+    @pytest.mark.tier1
+    @pytest.mark.parametrize("ocp_token_for_actor", [{"type": "free"}], indirect=True)
```
Comment (dbasunag) [inline tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py]: Same.

File: tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py (hunk 1 above)
Comment (SB159) [reply tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py]: Thanks, missed, Updated.

File: tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py (hunk 2 above)
Comment (SB159) [reply tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py]: Thanks, missed, Updated.

File: tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py [hunk 3]
```diff
@@ -63,6 +63,66 @@ def test_inference_model_in_body_only_invalid_key_returns_401(
...
+            json=streaming_payload,
+            timeout=60,
+            stream=True,
+        )
+        assert response.status_code == 200, f"Expected 200 for streaming BBR inference, got {response.status_code}"
+        content_type = response.headers.get("content-type", "")
+        assert "text/event-stream" in content_type, (
+            f"Expected text/event-stream content type for streaming response, got '{content_type}'"
+        )
+        chunks = [line for line, _ in zip(response.iter_lines(), range(10)) if line]
```
[the remark this answers is not in the document]
Comment (SB159) [reply tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py]: Addressed

File: tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py [hunk 4]
```diff
@@ -63,6 +63,66 @@ def test_inference_model_in_body_only_invalid_key_returns_401(
...
+        response = request_session_http.post(
+            url=bbr_inference_url,
+            headers=bbr_api_key_headers,
+            json=wrong_model_payload,
+            timeout=60,
+        )
+        assert response.status_code not in (401, 403), (
+            f"Auth rejected based on body model; expected path identity to win on /llm/ paths, "
+            f"got {response.status_code}"
+        )
```
[the remark this answers is not in the document]
Comment (SB159) [reply tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py]: Addressed

File: tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py (hunk 2 above)
Comment (dbasunag) [reply tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py]: This is not addressed.

File: tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py (hunk 1 above)
Comment (dbasunag) [reply tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py]: Same.

File: tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/conftest.py [hunk 5]
```diff
@@ -55,7 +55,7 @@ def bbr_inference_url(
     return f"{maas_scheme}://{maas_host}/llm/{maas_inference_service_tinyllama_free.name}/v1/chat/completions"
 
 
-@pytest.fixture(scope="class")
```
Comment (dbasunag) [inline tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/conftest.py:58]: Changing class scoped fixture would have unwanted side effect.

Comment (SB159) [reply tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/conftest.py:58]: Addressed in latest commit: class-level parametrize + class-scoped fixtures restored.

**Review by kpunwatk — APPROVED** [review]

File: tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py [hunk 6]
```diff
@@ -14,38 +14,23 @@
 
 LOGGER = structlog.get_logger(name=__name__)
 
+pytestmark = [
+    pytest.mark.usefixtures(
+        "maas_unprivileged_model_namespace",
+        "maas_subscription_controller_enabled_latest",
+        "maas_gateway_api",
+        "maas_api_gateway_reachable",
+        "maas_free_group",
```
Comment (sheltoncyril) [inline tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py:23]: **High confidence** — Moving `usefixtures` from class-level to module-level `pytestmark` applies these 10 fixtures to every test and class in the file. If someone adds a new test class or standalone function later, they silently inherit potentially expensive cluster setup fixtures. Consider keeping the decorator on each class explicitly for a narrower blast radius.

File: tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py [hunk 7]
```diff
@@ -14,38 +14,23 @@
...
+        "maas_model_tinyllama_free",
+        "maas_auth_policy_tinyllama_free",
+        "maas_subscription_tinyllama_free",
+        "maas_inference_service_tinyllama_free",
+    ),
```
Comment (sheltoncyril) [inline tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py:28]: **High confidence** — The old class had `@pytest.mark.parametrize("ocp_token_for_actor", [{"type": "free"}], indirect=True)` on each test method. After the split, `TestBBRPreAuthInferenceAuth` no longer has it — only `TestBBRPreAuthInference` does. If the `ocp_token_for_actor` fixture requires the indirect parametrize to be present, the auth tests (invalid key, no key) may error or pass for the wrong reason. Could you verify this is intentional?

File: tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py [hunk 8]
```diff
@@ -89,8 +74,84 @@ def test_inference_no_api_key_returns_401_or_403(
...
+        bbr_api_key_headers: dict[str, str],
+    ) -> None:
+        """Verify that inference with stream=True returns 200 with SSE text/event-stream chunks."""
+        streaming_payload = {**bbr_chat_payload, "stream": True}
+        with request_session_http.post(
+            url=bbr_inference_url,
+            headers=bbr_api_key_headers,
+            json=streaming_payload,
+            timeout=60,
+            stream=True,
```
Comment (sheltoncyril) [inline tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py:141]: **High confidence** — The `chunks` list can be empty if the server returns blank keep-alive lines before actual data (all filtered out by `if line`). In that case `any([])` returns `False` and the assertion fails with no diagnostic context. Consider adding `assert chunks, 'No SSE lines received from streaming response'` before the `data:` prefix check.

**Review by sheltoncyril — COMMENTED** [review]
## Review Summary (AI-assisted, multi-persona)

_This review was generated by an AI agent using three reviewer personas (chill, grumpy, unhinged) with confidence scoring. Findings flagged by multiple personas are higher confidence._

**High confidence (3):** streaming test empty-chunks edge case, module-level pytestmark blast radius, auth class missing ocp_token_for_actor parametrize.
**Medium confidence (4):** path-wins 404 assertion fragility, SSE data: prefix check too loose, status code assert doesn't gate chunk parsing, chunks variable scope after with-block.

File: tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py [hunk 9]
```diff
@@ -89,8 +74,84 @@ def test_inference_no_api_key_returns_401_or_403(
...
+
+    @pytest.mark.tier1
+    def test_inference_path_wins_over_body_model_for_auth(
+        self: Self,
+        request_session_http: requests.Session,
+        bbr_inference_url: str,
+        bbr_api_key_headers: dict[str, str],
+    ) -> None:
+        """Verify that on /llm/ paths, auth uses path identity — wrong body model returns 404, not 401/403.
+
```
Comment (sheltoncyril) [inline tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py:107]: **Medium confidence** — The test asserts `expected_status=404` for a wrong body model, but depending on server implementation this could also be 400 or 422. The test name says "path wins for auth" but a 404 proves routing behavior, not auth behavior. Consider clarifying the test name and adding a comment explaining why 404 specifically is the expected outcome in this setup.

File: tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py [hunk 10]
```diff
@@ -89,8 +74,84 @@ def test_inference_no_api_key_returns_401_or_403(
...
+        """Verify that inference with stream=True returns 200 with SSE text/event-stream chunks."""
+        streaming_payload = {**bbr_chat_payload, "stream": True}
+        with request_session_http.post(
+            url=bbr_inference_url,
+            headers=bbr_api_key_headers,
+            json=streaming_payload,
+            timeout=60,
+            stream=True,
+        ) as response:
+            assert response.status_code == 200, f"Expected 200 for streaming BBR inference, got {response.status_code}"
```
Comment (sheltoncyril) [inline tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py:143]: **Medium confidence** — `chunk.startswith(b"data:")` matches the SSE prefix but doesn't verify the payload structure. A malformed response like `data:garbage` (no space after colon) would also pass. Consider checking for `b"data: "` (with space per SSE spec), and optionally parsing at least one chunk as JSON to confirm the response contains valid inference data.

File: tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py [hunk 11]
```diff
@@ -89,8 +74,84 @@ def test_inference_no_api_key_returns_401_or_403(
...
+        LOGGER.info("Path-wins auth verified: auth passed and routing returned 404 for wrong body model")
+
+    @pytest.mark.tier1
+    def test_inference_streaming_returns_sse(
+        self: Self,
+        request_session_http: requests.Session,
+        bbr_inference_url: str,
+        bbr_chat_payload: dict[str, Any],
+        bbr_api_key_headers: dict[str, str],
+    ) -> None:
```
Comment (sheltoncyril) [inline tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py:133]: **Medium confidence** — The status code assertion is inside the `with` block, but if it fails, chunk parsing still proceeds on the error response body. A 500 error containing `data: internal error` could pass the SSE assertion. Consider returning early or skipping chunk parsing when the status code check fails.

File: tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py [hunk 12]
```diff
@@ -89,8 +74,84 @@ def test_inference_no_api_key_returns_401_or_403(
...
+            json=streaming_payload,
+            timeout=60,
+            stream=True,
+        ) as response:
+            assert response.status_code == 200, f"Expected 200 for streaming BBR inference, got {response.status_code}"
+            content_type = response.headers.get("content-type", "")
+            assert "text/event-stream" in content_type, (
+                f"Expected text/event-stream content type for streaming response, got '{content_type}'"
+            )
+            chunks = [line for line, _ in zip(response.iter_lines(), range(10), strict=False) if line]
```
Comment (sheltoncyril) [inline tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py:148]: **Medium confidence** — The `LOGGER.info` line after the `with` block references `chunks`, but if an assertion inside the `with` block raises before the list comprehension completes, `chunks` will be unbound, causing a secondary `NameError` that masks the real failure. Consider moving this log line inside the `with` block.

File: tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py (hunk 7 above)
Comment (sheltoncyril) [reply tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py:28]: **Resolved — false positive.** After checking the fixture chain locally: TestBBRPreAuthInferenceAuth tests (invalid key, no key) don't use bbr_api_key_headers or ocp_token_for_actor — they hardcode invalid/empty headers. The parametrize removal is intentional and correct for these negative auth paths.

File: tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py (hunk 9 above)
Comment (sheltoncyril) [reply tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py:107]: **Partially resolved.** The docstring (lines 106-110) explains why 404 is expected: BBR pre-auth sets X-Gateway-Model-Name from the body, so post-auth routing returns 404 for a mismatched model. Original comment missed this context. The naming suggestion still has minor merit but is not blocking.

File: tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py (hunk 11 above)
Comment (sheltoncyril) [reply tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py:133]: **Resolved — false positive.** The assert on line 143 raises AssertionError if status != 200, which stops execution before chunk parsing. Pytest assertions are exceptions — they do gate subsequent code. A 500 response would not reach the SSE assertions.

File: tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py (hunk 6 above)
Comment (dbasunag) [reply tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py:23]: They would be executing once per module, vs once per class. Since all the Maas fixtures this seems like a right move to me. I feel this is a false positive.

**Review by dbasunag — APPROVED** [review]

File: tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py (hunk 7 above)
Comment (SB159) [reply tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py:28]: Yes, intentional. Auth tests don’t need ocp_token_for_actor — they only assert 401/403 with invalid/missing credentials. Class-level parametrize is on TestBBRPreAuthInference only.

File: tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py (hunk 8 above)
Comment (SB159) [reply tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py:141]: Good point — edge case with blank keep-alive lines. Not seen in our runs; happy to add a clearer assert in a follow-up if needed.

File: tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py (hunk 9 above)
Comment (SB159) [reply tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py:107]: Docstring covers why 404 is expected (auth passes, routing fails for wrong body model). Haven’t seen 400/422 for this case; happy to adjust naming in a follow-up if needed.

File: tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py (hunk 10 above)
Comment (SB159) [reply tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py:143]: Intentionally minimal for smoke — just verify we get SSE data chunks. Happy to use data: or parse JSON if we need stricter validation later.

File: tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py (hunk 11 above)
Comment (SB159) [reply tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py:133]: With the current assert on status code, chunk parsing only runs after 200. Haven’t seen a false pass on error bodies

File: tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py (hunk 12 above)
Comment (SB159) [reply tests/model_serving/maas_billing/body_base_routing/pre_auth_model_header/test_bbr_pre_auth_inference.py:148]: Assert failures stop the test before that log line, so haven’t seen this in practice.

**Review by sheltoncyril — APPROVED** [review]
/lgtm
